extensions/tabs/scripts/confusion.js

```diff
@@ -0,0 +1,340 @@
+          .attr('cx', getXConfusionQuadrant(d, i))
+          .attr('cy', getYConfusionQuadrant(d, i))
+          .attr('stroke', getDotColor(d, i))
+          .attr('fill', getDotColor(d, i))
+          .attr('opacity', 0.5)
+          .classed('active', true)
+          .on('click', function () {
+            d3.selectAll('.clicked').classed('clicked', false)
+            var div = d3.select('#testData')
+            div.html(getTestResultHtml(d))
+            d3.selectAll('.active').classed('clicked', true)
+          })
+          .on('mouseover', function () {
+            tooltip.transition()
+              .duration(500)
+              .style('opacity', 0.9)
+            tooltip.html(d.score || 'not scored')
+              .style('left', (d3.event.pageX) + 'px')
+              .style('top', (d3.event.pageY - 28) + 'px')
+          })
+          .on('mouseout', function () {
+            tooltip.transition()
+              .duration(500)
+              .style('opacity', 0)
+            d3.selectAll('.active').classed('active', false)
+          })
+      })
+  }
+}
+
+function addGraph () {
+  var yAxis = d3.axisLeft().scale(yScale)
+
+  var svg = d3.select('#confusionMatrix').append('svg')
+    .attr('viewBox', '0, 0,' + outerWidth + ',' + outerHeight)
+    .append('g')
+    .attr('transform', 'translate(' + margin.left + ',' + margin.top + ')')
+
+  svg.append('rect')
+    .attr('class', 'outer')
+    .attr('width', innerWidth)
+    .attr('height', innerHeight)
+
+  var g = svg.append('g')
+    .attr('transform', 'translate(' + padding.left + ',' + padding.top + ')')
+
+  g.append('g')
+    .attr('class', 'y axis')
+    .attr('transform', 'translate(0,0)')
+    .call(yAxis)
+
+  g.selectAll('line.horizontalGrid').data(yScale.ticks(13))
+    .enter()
+    .append('line')
+    .attr('class', 'horizontalGrid')
+    .attr('x1', 0)
+    .attr('x2', width)
+    .attr('y1', function (d) { return yScale(d) })
+    .attr('y2', function (d) { return yScale(d) })
+    .attr('fill', 'none')
+    .attr('shape-rendering', 'crispEdges')
+    .attr('stroke', 'steelblue')
+    .attr('stroke-dasharray', '10,10')
+    .attr('stroke-width', '1px')
+
+  g.append('text')
+    .attr('x', 10)
+    .attr('y', 15)
+    .text('True Positive')
+  g.append('text')
+    .attr('x', width - 100)
+    .attr('y', 15)
+    .text('False Positive')
+  g.append('text')
+    .attr('x', 10)
+    .attr('y', height - 5)
+    .text('False Negative')
+  g.append('text')
+    .attr('x', width - 100)
+    .attr('y', height - 5)
+    .text('True Negative')
+  g.append('line')
+    .attr('x1', 0)
+    .attr('y1', height / 2)
+    .attr('x2', width)
+    .attr('y2', height / 2)
+    .attr('stroke', 'steelblue')
+  g.append('line')
+    .attr('x1', width / 2)
+    .attr('y1', 0)
+    .attr('x2', width / 2)
+    .attr('y2', height)
+    .attr('stroke', 'steelblue')
+}
+
+function getTestResultHtml (test) {
+  console.log('getTestResults')
+  var t = {
+    test: {
+      '<>': 'div',
```

Comment:
This kind of thing is exactly what JSX was made for (cough - React - cough)

Reply:
We'll get to this with react. ;)